src/Alfred/Cache.hs

```diff
@@ -1,11 +1,34 @@
 module Alfred.Cache
     ( writeItems
+    , rawWriteItems
     ) where
 
 import qualified Data.Csv as Csv
 import qualified Data.ByteString.Lazy as BS
 
 import Alfred.Item
+import Alfred.Search
+import qualified Data.HashMap.Strict as HM
 
 writeItems :: String -> [Item] -> IO ()
-writeItems cacheFile = BS.writeFile cacheFile . Csv.encodeDefaultOrderedByName
+writeItems cacheFile items = do
+  cachedItems <- getItems cacheFile
+  rawWriteItems cacheFile $ itemsWithCachedVisits items cachedItems
+
+rawWriteItems :: String -> [Item] -> IO ()
+rawWriteItems cacheFile = BS.writeFile cacheFile . Csv.encodeDefaultOrderedByName
+
+itemsWithCachedVisits :: (Foldable t) => [Item] -> t Item -> [Item]
+itemsWithCachedVisits currentItems itemsCache =
+  let
+    insert fn i = HM.insert (aiUrl i) (fn i)
+    lookup i = HM.lookup (aiUrl i)
+    toHash fn = foldr (insert fn) HM.empty
+    itemsCache' = toHash id itemsCache
+
+    itemWithCachedVisits i =
```

Comment:
Thoughts on pulling some of these functions out and giving them type signatures? Also, I think I prefer `where` over `let/in`. Allows you to see the high level stuff first.